vcd/resource_vcd_vapp_vm.go

```diff
@@ -497,6 +497,24 @@ var vappVmSchema = map[string]*schema.Schema{
 			},
 		},
 	},
+	"cpu_hot_add_enabled": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Default:     false,
+		Description: "True if the virtual machine supports addition of virtual CPUs while powered on.",
+	},
+	"memory_hot_add_enabled": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Default:     false,
+		Description: "True if the virtual machine supports addition of memory while powered on.",
+	},
+	"prevent_reboot": {
```

Comment:
Is this name coming from the API?

Reply:
it's not API thing, this is `sugar`. Idea created by Dainius. this allow user to stop vm restart if he don't know if update will happen in cold or hot manner.

Reply:
Interesting feature! But I was experimenting with it and it looks it needs more work. Issue:
1. Take a powered off VM.
2. Change/apply something in the VM through Terraform with `prevent_reboot = true`
3. The VM got powered back on!

Either we need to make sure it doesn't get powered on either, or we need two separate parameters, but that's an overkill, IMHO:
- `prevent_power_off = true`
- `prevent_power_on = true`

Reply:
Well reboot is power off and powen on. If you have state which is power off and make update and vm is powered on then it isn't restart. You can prevent power on by using:
* `power_on` - (Optional) A boolean value stating if this VM should be powered on. Default is `true`

Reply:
First, I confirm that `power_on` works as expected in this case. I.e. VM stays powered off:

```
# vcd_vapp_vm.demo_vm_photon will be updated in-place
  ~ resource "vcd_vapp_vm" "demo_vm_photon" {
...
      ~ cpu_hot_add_enabled            = false -> true
...
      ~ power_on                       = true -> false
}

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_vapp_vm.demo_vm_photon: Modifying... [id=urn:vcloud:vm:cdfccf69-ebbc-4918-b8c7-ad141776bedf]
vcd_vapp_vm.demo_vm_photon: Still modifying... [id=urn:vcloud:vm:cdfccf69-ebbc-4918-b8c7-ad141776bedf, 10s elapsed]
vcd_vapp_vm.demo_vm_photon: Modifications complete after 39s [id=urn:vcloud:vm:cdfccf69-ebbc-4918-b8c7-ad141776bedf]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

> Well reboot is power off and powen on. If you have state which is power off and make update and vm is powered on then it isn't restart. 

As you said, reboot is power off **and** on. Hence, current naming doesn't match actual behavior (that it doesn't prevent power on). I suggest renaming the parameter to reflect specifically to what it does to avoid possible confusion and to have a name that is "less smart" and you know instantly what it does:
- `prevent_update_power_off`

However, I still find this "sugar" not good for one's mental health. A proper approach would be:
1. To have a new action field `action_reboot = false`
2. Then, if someone is changing a parameter which requires a reboot, then this fields get set by Terraform to `action_reboot = true` and **is shown in the plan**.
3. Then the user can decide whether to proceed with update or not.

What do you think?

Reply:
changed to `prevent_update_power_off`

Reply:
> However, I still find this "sugar" not good for one's mental health. A proper approach would be:
> 
>     1. To have a new action field `action_reboot = false`
> 
>     2. Then, if someone is changing a parameter which requires a reboot, then this fields get set by Terraform to `action_reboot = true` and **is shown in the plan**.
> 
>     3. Then the user can decide whether to proceed with update or not.
> 
> 
> What do you think?

Interesting idea, but it won't work. We have to toggle this parameter true to false in one session - true for plan phase, back after update. User can provide this parameter too. We can't know if parameter is changed by us or others. So there are more questions when answers how it would work and how code should handle and if terraform will allow to change states back and for on read stages.

Reply:
> changed to prevent_update_power_off

Thank you, for me it's a lot clearer now (less fuzzy).

You mentioned that @Didainius suggested this, so CCing him.

> So there are more questions when answers how it would work and how code should handle and if terraform will allow to change states back and for on read stages.

I checked whether there's some solutions in the pipeline for Terraform 0.13, but it doesn't seem so:
https://www.hashicorp.com/blog/announcing-the-terraform-0-13-beta/

Reply:
> However, I still find this "sugar" not good for one's mental health. A proper approach would be:
> 
>     1. To have a new action field `action_reboot = false`
> 
>     2. Then, if someone is changing a parameter which requires a reboot, then this fields get set by Terraform to `action_reboot = true` and **is shown in the plan**.
> 
>     3. Then the user can decide whether to proceed with update or not.

Yeah,  I doubt that one can fit Terraform workflow. In theory we probably could throw a warning or error during `read`, but at `read` phase one can only access "state", not "new config" so far so this is a no-go yet. 

In general to me it is an ok approach (at least I don't know better) with main point being to give insurance for user that `apply` does not reboot servers (it can be quite tricky for user to know what can be altered online, and what can't).

Reply:
I think renaming to `prevent_update_power_off` made it specific/clear enough given available options. Resolving conversation.